File: src/cycle/clock.rs
//! Defines the `Clock` trait and its implementations for managing time in the system.
//!
//! This module provides a flexible way to handle time, allowing the system to run
//! with a simulated, step-by-step clock (for testing) or a real-time clock.

/// A trait for a time source.
pub trait Clock: Send + Sync {
    /// Returns the current time as a u64 timestamp.
    fn get_time(&self) -> u64;
    /// Advances the clock's state, if applicable.
    fn tick(&mut self);
}

/// A clock that increments a counter on each `tick`.
///
/// This is useful for simulations and testing, where time advances in discrete steps.
#[derive(Debug, Default)]
pub struct IterativeClock {
    time: u64,
}

impl IterativeClock {
    /// Creates a new `IterativeClock` starting at time 0.
    pub fn new() -> Self {
        IterativeClock { time: 0 }
    }
}

impl Clock for IterativeClock {
    fn get_time(&self) -> u64 {
        self.time
    }

    /// Advances the clock by one time unit.
    fn tick(&mut self) {
        self.time += 1;
    }
}

/// A clock that provides the current real-world time as a Unix timestamp.
///
/// This is used for real-time applications where the system needs to interact
/// with the external world.
#[derive(Debug, Default)]
pub struct UnixTimeClock;

impl UnixTimeClock {
    /// Creates a new `UnixTimeClock`.
    pub fn new() -> Self {
        UnixTimeClock
    }
}

impl Clock for UnixTimeClock {
    fn get_time(&self) -> u64 {
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_secs()
    }

    /// The `tick` method is a no-op for the real-time clock, as time
    /// advances automatically.
    fn tick(&mut self) {
        // No-op
    }
}

File: core/Clock.js
/**
 * A trait/interface for a time source.
 */
export class Clock {
  /**
   * Returns the current time as a number timestamp.
   * @returns {number} Current time
   */
  getTime() {
    throw new Error('getTime() must be implemented by subclass');
  }

  /**
   * Advances the clock's state, if applicable.
   */
  tick() {
    throw new Error('tick() must be implemented by subclass');
  }
}

/**
 * A clock that increments a counter on each tick.
 *
 * This is useful for simulations and testing, where time advances in discrete steps.
 */
export class IterativeClock extends Clock {
  constructor() {
    super();
    this.time = 0;
  }

  getTime() {
    return this.time;
  }

  /**
   * Advances the clock by one time unit.
   */
  tick() {
    this.time += 1;
  }
}

/**
 * A clock that provides the current real-world time as a Unix timestamp.
 *
 * This is used for real-time applications where the system needs to interact
 * with the external world.
 */
export class UnixTimeClock extends Clock {
  getTime() {
    return Math.floor(Date.now() / 1000); // Unix timestamp in seconds
  }

  /**
   * The tick method is a no-op for the real-time clock, as time advances automatically.
   */
  tick() {
    // No operation needed
  }
}

/**
 * A clock that provides millisecond precision time.
 *
 * This is useful for detailed timing in performance-critical scenarios.
 */
export class HighResolutionClock extends Clock {
  constructor() {
    super();
    this.startTime = performance.now();
    this.offset = Date.now();
  }

  getTime() {
    // Return time in milliseconds since start
    return Math.floor(this.offset + (performance.now() - this.startTime));
  }

  tick() {
    // No operation needed
  }
}